### DISCUSSION TITLE
Typed tags for Record CRUD API hooks

### DISCUSSION THREAD

**Discussion by FarazPatankar** [pocketbase/pocketbase#4011] [Q&A] [answered]
I've read through a bunch of discussions but I might be missing something obvious. I've followed the documentation and am trying to set up some hooks using JS/TS and can't quite find a way to add type-safety to the `tags` in the CRUD hooks. By default, they have the `string[]` type.

What I've done so far:
1. Run PB locally
2. Import collections from prod so the same DB structure exists
3. Reference `pb_data/types.d.ts`. Although I don't see any mention of my models here?

Am I missing something or is this not possible?

**Comment by ganigeorgiev** [comment, marked as answer]
We don't generate types for the user defined collections out of the box.
`pb_data/types.d.ts` are ambient types for the [JS hooks](https://pocketbase.io/jsvm/index.html) implementation.

**Reply by FarazPatankar** [reply]
Ah, I see. Any suggestions on how to achieve this?

I also have a separate question about an ideal local dev setup, should I just ask it here or would a separate thread make more sense?

**Reply by ganigeorgiev** [reply]
I'm sorry, I'm not sure what to recommend.

TypeScript support for the JS hooks is not planned for now (_I'm even starting to regret implementing the JS hooks due to the numerous users complains about their limitations, but will think what to do with them after https://github.com/pocketbase/pocketbase/discussions/3271_).

**Reply by FarazPatankar** [reply]
Oh, please don't apologize. You're doing great work with PB and any development/contribution IMO is appreciated.

I think there's a world where this can be achieved in combination with [pocketbase-typegen](https://github.com/patmood/pocketbase-typegen). I do wonder how user-friendly it'd be though.

**Reply by FarazPatankar** [reply]
Btw, unrelated but I noticed you don't have GH sponsors. Is there another way to support you/your work?

**Comment by yureckey** [comment]
> I'm sorry, I'm not sure what to recommend.
> 
> TypeScript support for the JS hooks is not planned for now (_I'm even starting to regret implementing the JS hooks due to the numerous users complains about their limitations, but will think what to do with them after #3271_).

I'm fairly new to PB, but I think JS hooks are a very strong feature, allowing to extend logic much more easily than with Go
